### PR TITLE
Add support for multistep synthetics API tests

### LINKED ISSUE #938: datadog_synthetics_test should support subtype 'multi', fails on import
A combined feature request / bug report, I suppose?

### Terraform Version

```
Terraform v0.14.4
+ provider registry.terraform.io/datadog/datadog v2.21.0
```

### Affected Resource(s)

-   datadog_synthetics_test

### Terraform Configuration Files

```hcl
resource "datadog_synthetics_test" "foo" {
  name    = "foo"
  message = "foo process is failing."
  type    = "api"

  locations = ["aws:eu-central-1"]

  options_list {
    tick_every = 1800
    retry {
      count    = 2
      interval = 300
    }
  }

  request_definition {}
  status  = "live"
  tags    = ["env:nonprod-1", "service:foo"]
}
```

### Expected Behavior

The resource should have been imported.

### Actual Behavior

An error was returned (I've filtered out irrelevant sensitive data):
```
module.datadog.datadog_synthetics_test.foo: Importing from ID "zq6-f97-xgq"...
module.datadog.datadog_synthetics_test.foo: Import prepared!
  Prepared datadog_synthetics_test for import
module.datadog.datadog_synthetics_test.foo: Refreshing state... [id=zq6-f97-xgq]

Error: error getting synthetics test: multi is not a valid SyntheticsTestDetailsSubType: {"status":"live","public_id":"zq6-f97-xgq","tags":["service:foo","team:bar","cluster_name:nonprod-1","env:nonprod"],"locations":["aws:eu-central-1"],"message":"foo process is failing.","name":"foo","monitor_id":29521526,"type":"api","subtype":"multi","config":{"steps":[{"name":"Get plans","request":{"url":"https://foo/bar","method":"GET","timeout":30,"basicAuth":{"username":"","password":""}},"subtype":"http","extractedValues":[],"id":"xp9-cqp-2er","assertions":[{"operator":"is","type":"statusCode","target":200}]},{"name":"Get app plan","request":{"url":"https://foo/bar","method":"GET","timeout":30,"basicAuth":{"username":"","password":""}},"subtype":"http","extractedValues":[],"id":"n23-jcj-bae","assertions":[{"operator":"is","type":"statusCode","target":200},{"operator":"contains","type":"body","target":"app_name"}]},{"name":"Resize app","request":{"url":"https://foo/bar","method":"POST","timeout":30,"basicAuth":{"username":"","password":""}},"subtype":"http","extractedValues":[{"parser":{"type":"json_path","value":"$.task_id"},"type":"http_body","name":"TASK_ID"}],"id":"fvt-3kv-psk","assertions":[{"operator":"lessThan","type":"responseTime","target":3000},{"operator":"is","type":"statusCode","target":202},{"operator":"is","property":"content-type","type":"header","target":"application/json"},{"operator":"contains","type":"body","target":"task_id"}]},{"name":"Task status","request":{"url":"https://foo/bar","method":"GET","timeout":30,"basicAuth":{"username":"=","password":"="}},"subtype":"http","extractedValues":[],"id":"m9q-ney-g4h","assertions":[{"operator":"lessThan","type":"responseTime","target":3000},{"operator":"is","type":"statusCode","target":200},{"operator":"is","property":"content-type","type":"header","target":"application/json"},{"operator":"doesNotContain","type":"body","target":"FAILURE"}]}],"configVariables":[]},"options":{"ci":{"executionRule":"skipped"},"retry":{"count":2,"interval":300},"min_location_failed":1,"monitor_options":{"notify_audit":false,"locked":false,"include_tags":true,"new_host_delay":300,"notify_no_data":false,"renotify_interval":0},"tick_every":900,"min_failure_duration":0}}
```

### Steps to Reproduce

```
terraform import module.datadog.datadog_synthetics_test.foo zq6-f97-xgq
```

### References

Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

-   https://github.com/DataDog/datadog-api-client-go/issues/678 seems related, but was closed.

### PR DESCRIPTION
**Pull request by romainberger**
Close #938

### CODE REVIEW

File: datadog/resource_datadog_synthetics_test_.go [hunk 1]
```diff
@@ -488,6 +509,87 @@ func syntheticsTestOptionsList() *schema.Schema {
...
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				"request_definition": {
+					Description:   "The request for the api step.",
+					ConflictsWith: []string{"request"},
```
Comment (zippolyte) [inline datadog/resource_datadog_synthetics_test_.go:577]:
```suggestion
```

File: datadog/resource_datadog_synthetics_test_.go [hunk 2]
```diff
@@ -1127,6 +1187,101 @@ func buildSyntheticsAPITestStruct(d *schema.ResourceData) *datadogV1.SyntheticsA
...
+			request.SetTimeout(float64(requestMap["timeout"].(int)))
+
+			requestHeaders := stepMap["request_headers"].(map[string]interface{})
+			if len(requestHeaders) > 0 {
+				request.SetHeaders(make(map[string]string))
+			}
+			for k, v := range requestHeaders {
+				request.GetHeaders()[k] = v.(string)
+			}
+
```
Comment (zippolyte) [inline datadog/resource_datadog_synthetics_test_.go:1221]: Let's build the map first, then set it.
Also we already know in advance the length of the map, so let's use that in the make.

File: datadog/resource_datadog_synthetics_test_.go [hunk 3]
```diff
@@ -1127,6 +1187,101 @@ func buildSyntheticsAPITestStruct(d *schema.ResourceData) *datadogV1.SyntheticsA
 
 	config.SetConfigVariables(configVariables)
 
+	if attr, ok := d.GetOk("api_step"); ok && syntheticsTest.GetSubtype() == "multi" {
```
Comment (zippolyte) [inline datadog/resource_datadog_synthetics_test_.go:1190]: all of this seems very similar to what's done above in the case of a single step test. Can we factorize it ?

**Review by zippolyte — CHANGES_REQUESTED** [review]

File: datadog/resource_datadog_synthetics_test_.go [hunk 4]
```diff
@@ -2008,6 +2257,71 @@ func updateSyntheticsAPITestLocalState(d *schema.ResourceData, syntheticsTest *d
...
+				}
+
+				localStep["request_client_certificate"] = []map[string][]map[string]string{localCertificate}
+			}
+
+			localSteps[i] = localStep
+		}
+
+		if err := d.Set("api_step", localSteps); err != nil {
+			fmt.Println("####################", err)
```
Comment (romainberger) [inline datadog/resource_datadog_synthetics_test_.go:2321]: Forgot to remove this for debug, I'll delete it

File: datadog/resource_datadog_synthetics_test_.go (hunk 3 above)
Comment (romainberger) [reply datadog/resource_datadog_synthetics_test_.go:1190]: Given that there are different models (`SyntheticsAPITest` / `SyntheticsAPIStep`) I'm not sure we can factorize everything? But I'll see if there are more fields that we can extract in functions.

**Review by zippolyte — APPROVED** [review]
LGTM

**Pull request merged**